clang/lib/AST/ExprConstant.cpp

```diff
@@ -12367,8 +12368,17 @@ bool IntExprEvaluator::VisitBuiltinCallExpr(const CallExpr *E,
                            BuiltinOp != Builtin::BI__lzcnt &&
                            BuiltinOp != Builtin::BI__lzcnt64;
 
-    if (ZeroIsUndefined && !Val)
-      return Error(E);
+    if (!Val) {
```

Comment:
Hmm...for these 2 type generic builtins, it might be nice to check the number of arguments first, then call `EvaluateInteger` on the correct arg, rather than calling `EvaluateInteger`, failing, then calling `EvaluateInteger` again on the second arg.

Reply:
`!Val` checks if the first argument is zero: https://github.com/llvm/llvm-project/blob/59bb4f259d78c7a1916462a1f6d12cf335523db1/llvm/include/llvm/ADT/APInt.h#L587

We always need to evaluate the first argument and check if it's zero, as the second argument is a fallback return value that's only used in that case.

Reply:
https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html#index-_005f_005fbuiltin_005fclzg

> Similar to __builtin_clz, except the argument is type-generic unsigned integer (standard, extended or bit-precise) and there is optional second argument with int type. No integral argument promotions are performed on the first argument. If two arguments are specified, and first argument is 0, the result is the second argument. If only one argument is specified and it is 0, the result is undefined.

The code as written is doing:
1. evaluate arg 0.
2. if evaluation is 0 and the builtin is clzg and num args > 1
3. evaluate arg 1

Consider that it may be cheaper to evaluate the number of args than the full subexpression (at least when the sub expression itself is more than just a literal iteger).

The code can be rewritten as:

```suggestion
  case Builtin::BI__builtin_clzg:
  case Builtin::BI__builtin_ctzg: {
    unsigned ArgToEval = E->getNumArgs() > 1 ? 1 : 0;
    APSInt Val;
     if (!EvaluateInteger(E->getArg(0), Val, Info))
      return false;
     if (E->getNumArgs() == 1 && !Val)
       return Error(E);
     return Success(...
```

Reply:
I'm not sure I understand. What would the `...` in `return Success(...` be? If the first argument is not zero then the builtin returns the leading/trailing zero count of that, even when called with 2 arguments, so we can't just `return Success(Val.count{l,r}_zero(), E);` on line 12379 of your suggested change as it would always return the second argument when the builtin is called with 2 arguments.

Reply:
ah, no sorry, I see now what you mean.

Reply:
Oh yeah I messed up my last reply after rewording part of it. The builtins only make use of the second argument if the first argument is zero, and they just return it directly and not its leading/trailing zero count.

Reply:
In that case:
```c++
case Builtin::BI__builtin_clzg:
case Builtin::BI__builtin_ctzg: {
  APSInt Val;
  if (!EvaluateInteger(E->getArg(0), Val, Info))
    return false;

  if (E->getNumArgs() == 1) {
    if (!Val)
      return Error(E);
    if (BuiltinOp == Builtin::BI__builtin_clzg)
      return Success(Val.countl_zero(), E);
    return Success(Val.countr_zero(), E);
  }

  if (!EvaluateInteger(E->getArg(1), Val, Info))
    return false;
  return Success(Val, E);

Reply:
though I guess that's basically what you have!